linear-algebra/src/main/scala/scalan/linalgebra/Matrices.scala

```diff
@@ -89,112 +101,234 @@ trait Matrices extends Vectors { self: MatricesDsl =>
     def *(matrix: Matr[T])(implicit n: Numeric[T], o: Overloaded1): Matr[T] = {
-      val mT = matrix.transpose
-      val res = rows.flatMap { row => (mT * row).items}
-      companion(res, matrix.numColumns)
+      def mT = matrix.transpose
+      def direct = rows.flatMap { row => (mT * row).items}
+      matrix match {
+        case DenseFlatMatrix(_, width) =>
+          DenseFlatMatrix(direct, width)
+        case CompoundMatrix(_, width) =>
+          DenseFlatMatrix(direct, width)
+        case ConstMatrix(value, width, _) =>
+          val rowsConstant = (reduceByRows *^ value).items.map(v => ConstVector(v, width))
+          CompoundMatrix(rowsConstant, width)
+        case DiagonalMatrix(diagonalValues) =>
+          val diagonalReplicated = Collection.replicate(numRows, diagonalValues).flatMap(coll => coll)
+          DenseFlatMatrix((DenseVector(rmValues) *^ DenseVector(diagonalReplicated)).items, numColumns)
+        case ConstDiagonalMatrix(diagonalValue, _) =>
+          DenseFlatMatrix((DenseVector(rmValues) *^ diagonalValue).items, numColumns)
+        // TODO: check case _
+        case _ => !!!("matcher for @matrix argument in DenseFlatMatrix.*(matrix: Matr[T]) is not specified.")
+      }
     }
 
-    def +^^(other: Matr[T])(implicit n: Numeric[T]): Matr[T] = {
-      companion((rows zip other.rows).flatMap { case Pair(v1, v2) => (v1 +^ v2).items }, numColumns)
+    @OverloadId("matrix")
+    def +^^(matrix: Matr[T])(implicit n: Numeric[T]): Matr[T] = {
+      matrix match {
+        case DenseFlatMatrix(rmValues1, _) =>
+          DenseFlatMatrix((DenseVector(rmValues) +^ DenseVector(rmValues1)).items, numColumns)
+        case CompoundMatrix(rows1, _) =>
+          DenseFlatMatrix((rows zip rows1).flatMap { case Pair(v1, v2) => (v1 +^ v2).items }, numColumns)
+        case ConstMatrix(value, _, _) =>
+          DenseFlatMatrix((DenseVector(rmValues) +^ value).items, numColumns)
+        case DiagonalMatrix(diagonal) =>
+          val width = diagonal.length
+          val mainDiagonalIndices = Collection.indexRange(width).map(i => i * width + i)
+          val newValues = (diagonal zip rmValues(mainDiagonalIndices)).map { case Pair(d, v) => d + v }
+          DenseFlatMatrix(rmValues.updateMany(mainDiagonalIndices, newValues), numColumns)
+        case ConstDiagonalMatrix(diagonalValue, width) =>
+          val mainDiagonalIndices = Collection.indexRange(width).map(i => i * width + i)
+          val newValues = rmValues(mainDiagonalIndices).map(v => diagonalValue + v)
+          DenseFlatMatrix(rmValues.updateMany(mainDiagonalIndices, newValues), numColumns)
+        // TODO: check case _
+        case _ => !!!("matcher for @matrix argument in DenseFlatMatrix.+^^(matrix: Matr[T]) is not specified.")
+      }
     }
 
     @OverloadId("matrix")
-    def *^^(other: Matr[T])(implicit n: Numeric[T]): Matr[T] = {
-      other match {
+    def *^^(matrix: Matr[T])(implicit n: Numeric[T]): Matr[T] = {
+      matrix match {
```

Comment:
Yes, it is not clear from the name. `x` or `y` actually better names, because `x` is something on the left, `y` - on the right.